### ISSUE TITLE
Allow multi-threaded access of a single Context in graal js

### ISSUE THREAD

**Issue by iitsoftware** [oracle/graal#2484]
Our product is event-driven that ensures single-thread access to a single Context. But there is one exception when the JS code registers a JS function as call back on an external library.

Example using a message listener on a MQTT consumer:

```
    var LISTENER = Java.extend(Java.type("org.eclipse.paho.client.mqttv3.IMqttMessageListener"), {
        messageArrived: function(topic, message) {
           // Register a callback to be executed in our event loop
            stream.executeCallback(function(msg) {
               // Do something
            }, message);
        }
    });
    this.client.subscribe(this.props["topicfilter"], qos(this.props["qos"]), new LISTENER());

```

When a message arrives, the 3rd party lib calls that listener which leads to the multi-thread access exception from Graal if the thread that registers the listener is still within the context. We ensure ourself that we don't get into race conditions by registering another call back that is executed in our event loop. But we cannot avoid that a Java lib calls into the JS code asynchronously. However, we can handle it!

This async receive is a common pattern in messaging. We cannot use a Java message listener as all our actions are initiated from the JS code.

Nashorn allows this. I suggest lifting the restrictions a bit by introducing a new option `polyglot.js.multithreadaccessrestricted` and set it to `true` by default. If I set it explicitly to `false`, I should handle it. Alternatively, make it part of Nashorn compatibility mode.

**Comment by iitsoftware** [comment]
May I ask if this issue is planned to be solved and in which upcoming release?

**Comment by chumer** [comment]
I think it is useful to link the Slack discussion here: 
https://graalvm.slack.com/archives/CP6RTC4LT/p1589981590103900

Could you provide more a little more information for this example?
1) Do you use this in an embedded context? Or do you use one of our language launchers? Do you use GraalVM or do you consume it by jar? 
2) Nashorn compatibility mode is enabled, do you want to move away or stay?
3) How do you ensure that parallel access is not happening? From the example it seems like event queue is running on a different thread than the current JavaScript application that schedules it. Could you show us with a bigger example, or pseudo code?

In the slack conversation I suggested three workarounds:

> 1) You can bind it to a Java host function instead. That should work.
> 2) You can also spawn new contexts for each thread.
> 3) You can leave the current context acquire a lock and then call into the guest function.

You responded with 

> 1) Our users have to stay within JS; they can’t use Java host functions.

Ok.

> 2) Impossible. How should I spawn a new Context from a JS function?

You can spawn a new context using Java API. Just like you access the other Java APIs.

> 3) Impossible. I don’t have access to the Context in JS and if I had, it would make the code Graal-specific.

Note that the Java <-> JavaScript interop behavior we are exposing is Graal specific. They just happen to be close to Nashorn, to ease migration. There is no standard or guarantee that this works with any other runtime.


> May I ask if this issue is planned to be solved and in which upcoming release?

Currently there is no feature scheduled for the next release. I don't think we have a solution, so it would be hard to say anything definitive.

Just to be clear. We stand by the decision to not allow parallel access to the same JS context. We elaborate this in great detail here [1]. But if the access is event queue based, so the context access is never actually in parallel then I agree it should be possible to use Graal.js. Other projects like ES4X [2] managed to implement an event queue based system on top of Graal.js.

[1] https://medium.com/graalvm/multi-threaded-java-javascript-language-interoperability-in-graalvm-2f19c1f9c37b
[2] https://github.com/reactiverse/es4x

**Comment by iitsoftware** [comment]
Here is an example from ES4X, RabbitMQ client, async receive:

```
// Create a stream of messages from a queue
client.basicConsumer("my.queue", (rabbitMQConsumerAsyncResult) => {
  if (rabbitMQConsumerAsyncResult.succeeded()) {
    console.log("RabbitMQ consumer created !");
    let mqConsumer = rabbitMQConsumerAsyncResult.result();
    mqConsumer.handler((message) => {
      console.log("Got message: " + message.body().toString());
    });
  } else {
    rabbitMQConsumerAsyncResult.cause().printStackTrace();
  }
});
```

This is a dedicated Vertx wrapper around the RabbitMQ client. It just calls the result handler from the Vertx event queue.

But we want to register a JS function directly as a callback on any async Java lib. We cannot write a wrapper for all Java libs our users use. I can register the function, no problem. But if the callback is activated while I'm still within the context, it fails. 

The only thing I want is a switch that allows me to put the result of an async callback into my event queue from the JS side. Or, alternatively, lock the context so that async libs have to wait before activating a callback.

**Comment by chumer** [comment]
I can still not see how you ensure that access is not in parallel. For me it looks like the callback can happen at any point during the execution of the parent JS code. You need to at least specify a point during the JS execution where the queue can be processed synchronously. Or maybe this needs better integration with es4x? In other words, the JS code needs to explicitly leave on the current thread so other threads can process. 

```
// Create a stream of messages from a queue
client.basicConsumer("my.queue", (rabbitMQConsumerAsyncResult) => {
  if (rabbitMQConsumerAsyncResult.succeeded()) {
    console.log("RabbitMQ consumer created !");
    let mqConsumer = rabbitMQConsumerAsyncResult.result();
    mqConsumer.handler((message) => {
      console.log("Got message: " + message.body().toString());
    });
  } else {
    rabbitMQConsumerAsyncResult.cause().printStackTrace();
  }
});
 client.processQueue();
```

The client.processQueue can then take the parent context explicitly leave and reenter. I think if you did that in Nashorn you were also not thread-safe as any JS code can access the global object.

**Comment by iitsoftware** [comment]
The point is this: Nashorn allows it and we are having hundreds of scripts doing it. But we do *know* that JS is single threaded and we handle it by immediately putting the result of the callback into our event queue. 

```
    var LISTENER = Java.extend(Java.type("org.eclipse.paho.client.mqttv3.IMqttMessageListener"), {
        messageArrived: function(topic, message) {
            stream.executeCallback(function(msg) {
                self.executeOutputLink("Out", msg);
            }, message);
        }
    });

```
This is save. Async functions must immediately enqueue with `stream.executeCallback` which puts it into our event queue.

**Comment by chumer** [comment]
> Async functions must immediately enqueue with stream.executeCallback which puts it into our event queue.

Did you try to write a generic version of IMqttMessageListener in Java? You can pass the guest function as value in the constructor. Then the notification of the listener does not require a guest context being entered, as no JS code is executed in parallel.

**Comment by iitsoftware** [comment]
The MQTT code was just an example. One of many libs users are using. I can't write a dedicated generic class for any lib.

Currently, GraalVM has a hard restriction on multi-threading which needs to be relaxed, otherwise you will exclude many use cases. Why not relax it under the umbrella of Nashorn compatibility?

**Comment by chumer** [comment]
> Currently, GraalVM has a hard restriction on multi-threading which needs to be relaxed, otherwise you will exclude many use cases. Why not relax it under the umbrella of Nashorn compatibility?

For us it does not make much difference whether it is Nashorn compatibility mode or not. We need to support and test all our use-cases. At the moment it is not even clear whether the example you are providing is thread-safe. Who says that a function declaration is not side-effecting? It is even worse if such capabilities are exposed to script developers and not system developers, who will have a very hard time arguing about thread-safety. Following the guidelines of Nashorn[1] to implement such accesses is just not good enough for us. Almost all safe use-cases described there are actually achievable in Graal.JS too, sometimes a little differently. 

One way to properly support your exact use-case (I think there are excellent workarounds) is to introduce the notion of side-effect free functions. We don't support this atm and this is quite a significant feature, so this might take a while to actually land (if ever).

[1] https://blogs.oracle.com/nashorn/nashorn-multithreading-and-mt-safety

**Comment by iitsoftware** [comment]
Which workarounds without calling Graal-specific code or dedicated class impl? I haven't found one.

**Comment by chumer** [comment]
It is a clean solution if the behavior is specified in the language and no encapsulation boundaries are broken. Java has multi-threading semantics and JavaScript has currently no such semantics,  that is why we require multi-threaded code to be written in Java. Any solution that would allow to go around that fact in an unsafe way would be pretty dirty imo.

**Comment by chumer** [comment]
> Which workarounds without calling Graal-specific code or dedicated class impl? I haven't found one.

Dedicated class impl should solve your problem. You can even load new jars from JS at runtime if you so desire.

**Comment by iitsoftware** [comment]
Clean / not clean discussions really don't help here. If you check the above Nashorn blog, he states they have to respect reality. And the reality is that you do not support this kind of pattern. This is sad as it is a complete show stopper for me. 

We have a microservice runtime where you can do everything out of the JS code. You can access any Java class without modification, without implementing any generic proxy classes in Java. This must run on Nashorn and Graal without modification and without any engine-specific calls. We have already done several modifications in our runtime to support Graal. But we can't do that in the JS scripts which are hundreds on customer sites.

I understand your point but as you already said in the Slack thread: This must actually be possible in Graal...

**Comment by chumer** [comment]
> I understand your point but as you already said in the Slack thread: This must actually be possible in Graal...

The quote is incomplete. I was saying it should be possible if you are not actually executing JS code in parallel. But you are actually executing JS code parallel.

> And the reality is that you do not support this kind of pattern. This is sad as it is a complete show stopper for me.

We definitely want to improve and support some safe parallelism in the future. But we want to do it a way we can maintain long term.

**Comment by iitsoftware** [comment]
This might do the trick:

[https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Proxy.html](https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Proxy.html)

Will need to test that.

**Comment by pmlopes** [comment]
The `es4x` example with rabbitmq works because of the internals of vert.x. Vert.x runs with multiple event loops, however once a event starts to be processed by one event loop, during all it's lifecycle we keep it "sticky" to that event loop, so the contexts are always right (most of the times).

We do have a couple of corner cases that it won't work for us. The simplest one is `vertx.executeBlocking()` where a user runs a callback that is known to call blocking code or consume a large amount of CPU time, in this case the callback is called from a thread pool (and like in this example it will fail).

Basically this is impossible to solve for us, because it order to comply to the requirements, we would need to have full control on the event loop from the worker thread pool. The job would need to block the event loop to run the callback and release it in the end, which in the end would be the same as just run the callback on the eventloop blocking it.

**Comment by iitsoftware** [comment]
We use vertx too (in another project). Although it is completely Java, the pattern is the same. We use the SwiftMQ client and an async JMS message listener. What we do in `onMessage` is to use `vertx.runOnContext(...)` to process the received messages out of the vertx event queue. This works fine and nobody complains about non-clean code. So it's a common pattern. However, running vertx from JS on Graal this would fail...

Anyways, let me try to use the `Proxy` class. May be that is a workaround.

**Comment by pmlopes** [comment]
I'm not sure how `Proxy` would help, but share your results. Taking vert.x example, the js context is bound to the event loop, you pass a callback which refers to the context to another thread (in the worker thread pool), the moment the proxy invokes the underlying js function, you're still on the wrong thread.

Even if you would manage to create a new context for each thread on the worker thread pool, the original callback function would still refer to objects on the source (event loop) thread and you would again fail i think.

The only approach (in vert.x) to avoid this is to use the eventbus (yet there are well known bugs needed to be solved on vert.x to make it fully work on all cases). In this case in the event loop you send a message (the message must be serialized so it can be decouple of the event loop using the custom codec we have in `es4x` (which only works for a few cases). Then the message is passed to the worker thread pool and can be used at will, switching threads, blocking, etc... in the end the reply is passed back to the event loop, which deserializes the message back to a js object on the event loop context and you are safe.

In order to properly solve this there are a couple of things still to be done on vert.x:

1. in `es4x` we need to properly serialize js objects (which depend of a feature that was merged/fixed after the 20.1.0 release)
2. in `vertx-core` we need to make the eventbus encode/decode not lazy (to ensure we are not on the wrong thread

I've plans to do this, but I can't promise any ETA as atm we're trying to finish the 4.0.0 release.

For more info on this see:

https://groups.google.com/forum/#!topic/vertx-dev/7hI3M8I5hNA
https://github.com/reactiverse/es4x/issues/342

**Comment by iitsoftware** [comment]
It seems I found a solution with the `Proxy` class. Here is my `AsyncProxy` (just testing, needs to be cleaned):

```
package com.swiftmq.impl.streams;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class AsyncProxy implements java.lang.reflect.InvocationHandler {

    private Object obj;
    private Stream stream;

    public static Object newInstance(Stream stream, Object obj) {
        return java.lang.reflect.Proxy.newProxyInstance(
                obj.getClass().getClassLoader(),
                obj.getClass().getInterfaces(),
                new AsyncProxy(stream, obj));
    }

    private AsyncProxy(Stream stream, Object obj) {
        this.stream = stream;
        this.obj = obj;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        stream.executeCallback(new FunctionCallback() {
            @Override
            public void execute(Object context) {
                try {
                    method.invoke(obj, args);
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                }
            }
        }, null);
        return null;
    }
}

```

Here is the JS part:

```
    var self = this;
    var ASYNC_PROXY = Java.type("com.swiftmq.impl.streams.AsyncProxy");
    var LISTENER = Java.extend(Java.type("org.eclipse.paho.client.mqttv3.IMqttMessageListener"), {
        messageArrived: function(topic, message) {
            self.executeOutputLink("Out", message);
        }
    });
    this.client = this.getInputReference("Connection")().getClient();
    this.client.subscribe(this.props["topicfilter"], qos(this.props["qos"]), ASYNC_PROXY.newInstance(stream, new LISTENER()));

```

What it does is to execute the listener from our event queue (via `stream.executeCallback`). Works great.

**Comment by pmlopes** [comment]
Yes I see that it works as you run the callback back on the event loop. Just be careful, because if it is an event loop, and the callback is doing lots of CPU time, it will block it. I don't know what side effects that has in `paho`

**Comment by iitsoftware** [comment]
Right, but all I wanted is to register an async callback on any 3rd party Java lib without violating the JS single-threaded policy. This can now be done quite elegant with `Proxy`. I'll close the issue.

**Comment by iitsoftware** [comment]
Well, after some more testing it turns out that a `Proxy` doesn't work either. I'm getting this `InvalidStateException` while I register a callback on an async method when the callback is called while I'm still in that context.

I can’t use dedicated contexts per thread as I’m using thread pools that drive the event loops per script. Multiple contexts would require multiple `eval()` calls either. So, no way.

Proposal to fix that and make it Nashorn compliant:

- Establish a lock on `context.enter` and release it on `context.leave`. This would synchronize access to the context and you have it.
- Disable the checks through a system property

**Comment by iitsoftware** [comment]
Found a working solution. Examples can be found here: https://github.com/iitsoftware/graaljs-concurrency-problem
